src/Uno.UWP/Security/Authentication/Web/WebAuthenticationBrokerProvider.Android.cs

```diff
@@ -60,62 +65,165 @@ static IEnumerable<string> ExtractSchemes(IntentFilterAttribute a)
+					return false;
+				}
 
-			Interlocked.Exchange(ref _completionSource, null)
-				?.TrySetResult(new WebAuthenticationResult(null, 0, WebAuthenticationStatus.UserCancel));
+				tcsResponse?.TrySetResult(new WebAuthenticationResult(intentUri.OriginalString, 0, WebAuthenticationStatus.Success));
+				return true;
+			}
+			catch (Exception ex)
+			{
+				tcsResponse.TrySetException(ex);
+				return false;
+			}
 		}
 
-		private static string? _waitingForCallbackUri;
-
-		private static TaskCompletionSource<WebAuthenticationResult>? _completionSource;
-
 		protected virtual async Task<WebAuthenticationResult> AuthenticateAsyncCore(
 			WebAuthenticationOptions options,
 			Uri requestUri,
 			Uri callbackUri,
 			CancellationToken ct)
 		{
-			var tcs = new TaskCompletionSource<WebAuthenticationResult>();
-			Interlocked.Exchange(ref _completionSource, tcs)?.TrySetCanceled();
+			var url = requestUri;
+			var callbackUrl = callbackUri;
+			var packageName = Application.Context.PackageName;
 
-			_waitingForCallbackUri = callbackUri.OriginalString;
+			// Create an intent to see if the app developer wired up the callback activity correctly
+			var intent = new Intent(Intent.ActionView);
+			intent.AddCategory(Intent.CategoryBrowsable);
+			intent.AddCategory(Intent.CategoryDefault);
+			intent.SetPackage(packageName);
+			intent.SetData(global::Android.Net.Uri.Parse(callbackUrl.OriginalString));
 
-			await LaunchBrowserCore(options, requestUri, callbackUri, ct);
+			// Try to find the activity for the callback intent
+			if (!IsIntentSupported(intent, packageName))
+				throw new InvalidOperationException($"You must subclass the `{nameof(WebAuthenticationBrokerActivityBase)}` and create an IntentFilter for it which matches your `{nameof(callbackUrl)}`.");
 
-			var result = await tcs!.Task;
+			// Cancel any previous task that's still pending
+			if (tcsResponse?.Task != null && !tcsResponse.Task.IsCompleted)
+				tcsResponse.TrySetCanceled();
 
-			if(Interlocked.CompareExchange(ref _completionSource, null!, tcs) == tcs)
+			tcsResponse = new TaskCompletionSource<WebAuthenticationResult>();
+			currentRedirectUri = callbackUrl;
+
+			if (!(await StartCustomTabsActivity(url)))
 			{
-				_waitingForCallbackUri = null;
+				// Fall back to opening the system-registered browser if necessary
+				var urlOriginalString = url.OriginalString;
+				var browserIntent = new Intent(Intent.ActionView, global::Android.Net.Uri.Parse(urlOriginalString));
+				WebAuthenticationBrokerRedirectActivity.StartActivity(CurrentActivity!, browserIntent);
 			}
 
-			return result;
+			return await tcsResponse.Task;
 		}
 
-		protected virtual async Task LaunchBrowserCore(WebAuthenticationOptions options,
-			Uri requestUri,
-			Uri callbackUri,
-			CancellationToken ct)
+		private async Task<bool> StartCustomTabsActivity(Uri url)
```

Comment:
should you forward the `CancellationToken` here? (same in `BindServiceAsync`)

Reply:
yes, done